hw/mcu/nordic/nrf52xxx/src/hal_trng.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include "mcu/cmsis_nvic.h"
+#include "mcu/nrf52_hal.h"
+#include "syscfg/syscfg.h"
+#include "hal/hal_trng.h"
+
+#define min(a, b) ((a) < (b) ? (a) : (b))
+
+static uint8_t rng_cache[ MYNEWT_VAL(TRNG_CACHE_LEN) ];
+static uint16_t rng_cache_head;
+static uint16_t rng_cache_tail;
+
+static void
+nrf52_rng_start(void)
+{
+    int sr;
+
+    __HAL_DISABLE_INTERRUPTS(sr);
+
+    NRF_RNG->EVENTS_VALRDY = 0;
+    NRF_RNG->INTENSET = 1;
+    NRF_RNG->TASKS_START = 1;
+
+    __HAL_ENABLE_INTERRUPTS(sr);
+}
+
+static void
+nrf52_rng_stop(void)
+{
+    int sr;
+
+    __HAL_DISABLE_INTERRUPTS(sr);
+
+    NRF_RNG->INTENCLR = 1;
+    NRF_RNG->TASKS_STOP = 1;
+    NRF_RNG->EVENTS_VALRDY = 0;
+
+    __HAL_ENABLE_INTERRUPTS(sr);
+}
+
+static void
+nrf52_rng_irq_handler(void)
+{
+    if (NRF_RNG->EVENTS_VALRDY) {
+        NRF_RNG->EVENTS_VALRDY = 0;
+        rng_cache[rng_cache_tail] = NRF_RNG->VALUE;
+        rng_cache_tail++;
+        if (rng_cache_tail >= sizeof(rng_cache)) {
+            rng_cache_tail = 0;
+        }
+    }
+
+    NRF_RNG->INTENCLR = 1;
+    NRF_RNG->EVENTS_VALRDY = 0;
```

Comment:
This bit was already cleared above, inside the `if`, should it really be cleared again here?

Reply:
no need to :)